compiler/rustc_mir_transform/src/const_goto.rs

```diff
@@ -28,7 +28,7 @@ pub struct ConstGoto;
 
 impl<'tcx> MirPass<'tcx> for ConstGoto {
     fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
-        sess.mir_opt_level() >= 4
+        sess.mir_opt_level() >= 2
```

Comment:
It seems like DelayConstAssignments + ConstProp + SimplifyCfg ought to be able to handle everything that ConstGoto used to. (Which would be good, because those optimizations are more general than ConstGoto.)

e.g. for the simple example in ConstGoto's header:
```
bb2: {
    _2 = const true;
    goto -> bb3;
}
bb3: {
    switchInt(_2) -> [false: bb4, otherwise: bb5];
}
```
I'd expect DelayConstAssignments to change it to
```
bb2: {
    goto -> bb3;
}
bb3: {
    _2 = const true;
    switchInt(_2) -> [false: bb4, otherwise: bb5];
}
```
which ConstProp/SimplifyCfg should be able to handle.

Do you know of any situations where this doesn't work?

Reply:
In case it's not obvious I have no formal training or education of any kind in compilers or optimizations. I'm just looking through the MIR that comes out of the compiler today for things that look derpy and trying to fix them.

I think before ConstProp we can have code like this:
```
bb0: {
    _8 = const 1;
    _9 = const 2;
    _10 = Add(move _8, _move 9);
    switchInt(move _3) -> [0: bb1, 1: bb3, otherwise: bb2];
}
```
And here I am not sure the amount of analysis we'd need to figure out if _10 is being assigned a const is simpler than just doing ConstProp.

Reply:
Aside from the naming, it wasn't obvious. This pass is sensible :)

I may not have been clear--I think ~we should be able to remove ConstGoto (not ConstProp) after this change, because SinkConstAssignments should move the const assignment right into the block with the `switchInt`, which ConstProp/SimplifyCfg can handle without ConstGoto. Can you try your motivating `Option::map` example without ConstGoto, and see if it still gets optimized?~ Ah, I thought ConstGoto was more limited than it is--it allows multiple predecessors, so it can't be replaced like that. This is fine as-is then. (FWIW, the optimization that ConstGoto performs is referred to elsewhere as jump threading.)

It's a bit unfortunate that DataflowConstProp doesn't help here. It would know the value of the constant in the predecessor block, but there aren't any usages in the predecessor, so it doesn't do anything that would help ConstGoto. Maybe if it was available to other passes as an analysis...